package/base-files/files/lib/functions/ipv4.sh

```diff
@@ -0,0 +1,166 @@
+uint_max=4294967295
+
+# Common error handler
+__die() {
+    echo "$1" >&2
+    exit 1
+}
+
+# check that $1 is only base 10 digits, and that it doesn't
+# exceed 2^32-1
+check_uint32() {
+    local n="$1"
+
+    [ -z "$n" -o -n "${n//[0-9]/}" ] && __die "Not a decimal integer ($n)"
```

Comment:
For prefixes, awk's `int` was used which only accepts decimals - and it's reasonable to carry over.
But for doing calculations, it is reasonable to write integer addresses using the hexadecimal notation.
This would apply to `int2ip`, `is_contiguous`, and `netmask2prefix`.

Reply:
It is easy enough to use shell perform the conversion `$((ip))` while invoking the function.
Still, this would bug me less if it were named to describe more clearly what it's doing.
`check_decimal_uint32` would be an appropriate, imo.

I could also see two versions existing: `require_decimal_uint32` and `convert_to_decimal_uint32` (names to point out difference).

Linux tools tend to be fairly relaxed with the other numeral systems: `ip a add 10.0xa.0.012/0x18 ..`.
No idea, though, how the landscape looks like in Openwrt.

Reply:
> For prefixes, awk's `int` was used which only accepts decimals - and it's reasonable to carry over. But for doing calculations, it is reasonable to write integer addresses using the hexadecimal notation. This would apply to `int2ip`, `is_contiguous`, and `netmask2prefix`.

If I'm doing a calculation, then I can use `$((0xffff0000))` to do the base conversion for me.

Reply:
> It is easy enough to use shell perform the conversion `$((ip))` while invoking the function. Still, this would bug me less if it were named to describe more clearly what it's doing. `check_decimal_uint32` would be an appropriate, imo.

We're getting into bike-shedding territory.

If the names become too long, then the calling code will start to look ungainly.

I say we put it out there and see what the reactions are.

Reply:
I've renamed it `assert_uint32` and it returns true/false, and if it fails, one should `exit 1`.

Reply:
Look, we both want this merged and I don't want to contribute to any misery. You don't have to change anything to please me. You are much more 'Openwrt' than I am.
You asked me for a review and I'm offering some time of my life for feedback.

Assertions are generally expected to end the current control flow, which is why I proposed to include the concept in the name - people should at least get a chance to learn that their program can halt.
Changing the name to `assert_` and going from exit to return values at the same time is nonsensical from that perspective. Now, people can within a reasonable mind expect that their control flow will be halted for illegal values when, in reality, it won't.

I've also pointed out that passing the argument to `check_uint32` through shell `$((ip))`. The point is that, unless your look how `check_uint32` is implemented, you don't know you need to be doing that.
The same goes for the interrupted control flow. That is not bike-shedding - it's API design 101.

`check_uint32` communicates 'I can call this to check if my integer is in range'. It doesn't communicate numeral system requirements, it doesn't communicate that the ended control flow, it doesn't communicate the usage of stdout.

What you would intuitively want to with `check_uint32` do is something like this:

```shell
if ! check_uint32 "$inside_tunnel_netmask"; then
  echo "server-supplied netmask is invalid: $inside_tunnel_netmask" >&2
  continue
fi
```

In it's current form, I would only be getting `Not normalized notation` and have to go looking where the number came from.